### ISSUE TITLE
issue: Issues with `touchedFields` typing after upgrading to TypeScript 5.4

### ISSUE THREAD

**Issue by sashamorozov** [react-hook-form/react-hook-form#11603]
### Version Number

7.51.0

### Codesandbox/Expo snack

https://www.typescriptlang.org/play?ts=5.4.2#code/JYWwDg9gTgLgBDAnmApnA3nAYtEBlGAQxhQBptgUAbAEwAViALcrS2gNUKoFcUBnOAF84AMygQQcAERQUhAMYwAtIwgQA1kpG4pAbgBQ++RAB2feAHMUMAKJRxUOAF44AHgAqrajU49+cFAAPEhMaAS8OLl4+ck82GgA5QhA0IJCwim8GGEYPCJ8o-gA+IoAKfThRXAAubFwCYhQ8+N9ootIKuBNklFq47ySU-QBKZyKMTuAROFKAQlntKBAAOhR7aD4AbW6UgF04ADIDmfnFlZgIbnlGFBp8rZ2UfYAfZ6ql5eA+PG4AIxBgDASDRhqN0J1KrIYNwoCZ3is1g4Hj1dgB+ZYpPh8QhWTqCTpQmFw7ihFAiYAmW76QS6IA

### Steps to reproduce

Hello,

I've encountered an issue with the typing of the `touchedFields` property in `react-hook-form` after updating to TypeScript 5.4. It seems that the new version of TypeScript has changed the way types are interpreted or handled, causing the types related to `touchedFields` to now throw compilation errors.

A code example demonstrating the issue is available in [TypeScript Sandbox](https://www.typescriptlang.org/play?ts=5.4.2#code/JYWwDg9gTgLgBDAnmApnA3nAYtEBlGAQxhQBptgUAbAEwAViALcrS2gNUKoFcUBnOAF84AMygQQcAERQUhAMYwAtIwgQA1kpG4pAbgBQ++RAB2feAHMUMAKJRxUOAF44AHgAqrajU49+cFAAPEhMaAS8OLl4+ck82GgA5QhA0IJCwim8GGEYPCJ8o-gA+IoAKfThRXAAubFwCYhQ8+N9ootIKuBNklFq47ySU-QBKZyKMTuAROFKAQlntKBAAOhR7aD4AbW6UgF04ADIDmfnFlZgIbnlGFBp8rZ2UfYAfZ6ql5eA+PG4AIxBgDASDRhqN0J1KrIYNwoCZ3is1g4Hj1dgB+ZYpPh8QhWTqCTpQmFw7ihFAiYAmW76QS6IA)

TS Error: `Unsafe assignment of an "any" value.`

It would be greatly appreciated if you could look into this issue and suggest a solution or workaround.

Thank you in advance for your attention to this matter. I look forward to your response.

Best regards,
Aleksandr

### Expected behaviour

After upgrading to TypeScript 5.4, the `touchedFields` property should continue to work without any typing issues

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
This breaking change is coming from TS not this lib.

**Comment by hybs123** [comment]
TFieldName is extracting the names of the touched fields as string but somehow TS is not able to index touchedFields in the form.

**Comment by VictorGosse** [comment]
Hello,
I confirm that we are facing the exact same issue and I'm wondering if it might come from https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#more-aggressive-reduction-of-intersections-between-type-variables-and-primitive-types somehow.

**Comment by Lexe003** [comment]
Same issue here with `dirtyFields`

**Comment by ultrox** [comment]
![CleanShot 2024-05-19 at 10 00 22](https://github.com/react-hook-form/react-hook-form/assets/3077558/b60f73ce-71ea-4423-9f42-815830a695af)

`tsc -b` with Version 5.4.5

**Comment by denniske** [comment]
I see that there is a newer typescript version 5.6.3 now in the package.json of this library. Does this mean this issues is fixed?

**Comment by MikyStar** [comment]
Hi everyone, any news on that matter ? I'm also experiencing this issue when trying to index dirtyFields

**Comment by alexw-at** [comment]
I ran into the same issue and came up with this workaround

https://www.typescriptlang.org/play/?ts=5.4.2#code/JYWwDg9gTgLgBDAnmApnA3nAYtEBlGAQxhQBptgUAbAEwDVCqBXFAZzgF84AzKCEOACIoKQgGMYAWgAWECAGtJ3XIIDcAKHViIAO1bwA5ihgBRKHyhwAvHAA8AFSyVaDZmzgoAHiR012T6npGFlYAPgAKdTgeXAAubFwCYhQHAJdgtlDSKLgdQhAUePkURAhuBKh8IhJU5yC3MIBtAHIYCCYxaRQaNL9mgF04ADI4YtLynEqkmsc61xDQlpRzaFYB9QBKa1CMHOBy8IBCQ+VKgDpli1ZGvILBoZGjk9wzto6unrrr25RBgB8-jFzsBWHgmAAjEDAGAkGgbLboHLREQwJhQHRAkAXFZQb75X4AfjOBVYrEIRhyHByKLRGKYvhQ3GAOm66g4qiAA

**Comment by ap0nia** [comment]
This is correct, `FieldPath` is not necessarily a valid key. `form.touchedFields` is a nested object, but the `name` is a dot-concatenated string.

A `FieldPath` name might be `a.b.c`, but that is not actually a valid key because the original object was `{ a: { b: { c } }`. `touchedFields` would follow the shape of the "original" object, while `name` represents a path to index into this object, e.g. with the `get` utility function.

```ts
const name = "a.b.c"
const value = { a: { b: {  c: true } } }

// Invalid.
value[name]
value["a.b.c"]
```

See below, as well as the [playground](https://www.typescriptlang.org/play/?ts=5.4.2#code/JYWwDg9gTgLgBDAnmApnA3nAYtEBlGAQxhQBptgUAbAEwAViALcrS2gNUKoFcUBnOAF84AMygQQcAERQUhAMYwAtIwgQA1kpG4pAbgBQ+pKjiceaALwZ9cOIQBc127YBGj9Dee35jvjCjAAHYA5p62gp4REUbIaAByhCCWFNT0TAA8ZrwAfPoA9HnOAHoA-IbGaAAqENzyjCg0cFY4UPhEJJlcOQDaAOQwNXUNrKl8vQC6+YW2pYZUKPADtfU0jtXLDU0YwoQChIGIhgVwcfwkjRAuAFYoinwAdEaDK-eEJfcu7-JzC3CBiShHAkklteoQPvd5L0joVKoxgAIaBB+H8IPAAO7QdSPJZDGjdf5JcZAA).

```ts
import type { FormState, FieldPath, FieldValues } from "react-hook-form";

type Value = {
  a: {
    b: {
      c: string
    }
  }
}

type Name = FieldPath<Value>
//   ^? "a" | "a.b" | "a.b.c"

type Touched = FormState<Value>['touchedFields']
//   ^? { readonly a?: { b?: { c?: boolean } } }

let touched: Touched = {} as any

// Nested objects.
touched.a?.b?.c

let name: Name = 'a.b.c'

// @ts-expect-error This does not work because there is no such property "a.b.c".
touched[name]
```

## "Solution"

The proper way to use a `FieldPath` is with the `get` utility. You may need to manually define the type if you don't like the default `any` returned by the function.

See the [playground example](https://www.typescriptlang.org/play/?ts=5.4.2&ssl=12&ssc=3&pln=1&pc=1#code/JYWwDg9gTgLgBAbzgcwKbwL5wGZQiOAIilQEMBjGAWgAsIIBrK7aEQgbgChRJY4YAnmFSI4AMVYBlGKRioANOOCoANgBMACrJqKxy9QDVSKgK6oAznCy58REhWp1GzVh06dyEAHbn4aGACiUHhQcAC8cAA8ACp6qmpGphZwqAAecl5qlnGGxmbmirH6agBypCAiaRlZSvFaMDQxOQl5FgB8bQAUnHA4rABc4lIyck3Fiflt8j1wXuWog0XxZRWcAJThbYgzwNhwnQCEByxQIAB0qMHQ5gDacxUAunAAZM-7R-6dJ+cwECbkNFQamaBXuqA2AB8IX1TmdgOZJCYAEYgYAwORqNYbBAzXokGAmKBeGHnS4hW5gh4AfjOFXM5lIaBmGBm+MJxJMmVQ2GAXiBnAw7CAA)

```ts
import { get } from "react-hook-form";
import type { FormState, FieldPath, FieldValues } from "react-hook-form";

const getError = <TFieldValues extends FieldValues, TFieldName extends FieldPath<TFieldValues>>(
  form: FormState<TFieldValues>,
  name: TFieldName
) => {
  if (!!form.errors[name] && (!!get(form.touchedFields, name) || form.isSubmitted)) {
    return form.errors[name]?.message
  }
  return undefined
};
```
